File: app/models/_report.py
"""Quality-report domain model + write-time validation.

Persisted as a single YAML file at ``<project>/report/<file_name>.yaml`` —
see ``specs/features/12-feature-quality-report-NEW.md``. Pure (no FS, no
HTTP).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from ..errors import ValidationError
from ._common import (
    ENUM_IDENTIFIER_RE,
    REPORT_TYPES,
    RUN_RESULTS,
    RUN_SET_TYPES,
    _is_single_line,
)


@dataclass(slots=True)
class Report:
    """A persisted quality report.

    Stored as a single YAML file at ``<project>/report/<file_name>.yaml``.
    ``type`` is the immutable discriminator (:data:`REPORT_TYPES`); only the
    config fields relevant to that type are populated. Run-set types carry
    ``run_paths`` (editable); the folder type (``tag_inventory``) carries a
    ``scope`` + surveyed ``tag``. All paths are data-root-relative POSIX and
    include the project, consistent with :class:`RunResult` / :class:`TestRun`.
    """

    type: str = ""
    title: str = ""
    created_at: str = ""
    run_paths: list[str] = field(default_factory=list)
    scope: str = ""
    status: str = ""
    kind: str = ""
    case_path: str = ""
    tag: str = ""

    def to_dict(self) -> dict[str, Any]:
        return {
            "type": self.type,
            "title": self.title,
            "created_at": self.created_at,
            "run_paths": list(self.run_paths),
            "scope": self.scope,
            "status": self.status,
            "kind": self.kind,
            "case_path": self.case_path,
            "tag": self.tag,
        }

    @classmethod
    def from_dict(cls, payload: dict[str, Any]) -> "Report":
        return cls(
            type=str(payload.get("type", "")),
            title=str(payload.get("title", "")),
            created_at=str(payload.get("created_at", "")),
            run_paths=[str(p) for p in payload.get("run_paths", [])],
            scope=str(payload.get("scope", "")),
            status=str(payload.get("status", "")),
            kind=str(payload.get("kind", "")),
            case_path=str(payload.get("case_path", "")),
            tag=str(payload.get("tag", "")),
        )


def validate_report(report: Report) -> None:
    """Raise :class:`ValidationError` if ``report`` violates pure invariants.

    Disk/project cross-checks (does ``kind`` exist in ``enums.yaml``? do the
    ``run_paths`` resolve? is ``case_path`` / ``scope`` a real path?) are
    storage concerns, not model concerns — mirroring :func:`validate_feature`
    vs the storage enum cross-check in spec 11.
    """

    if report.type not in REPORT_TYPES:
        raise ValidationError(
            field="type",
            message=(
                f"Invalid report type: {report.type!r}. "
                f"Must be one of {sorted(REPORT_TYPES)}."
            ),
        )

    if not report.title.strip():
        raise ValidationError(
            field="title", message="Report title must not be empty."
        )
    if not _is_single_line(report.title):
        raise ValidationError(
            field="title", message="Report title must be single-line."
        )

    if not report.created_at.strip():
        raise ValidationError(
            field="created_at", message="created_at must not be empty."
        )
    if not _is_single_line(report.created_at):
        raise ValidationError(
            field="created_at", message="created_at must be single-line."
        )

    # -- Per-type config -------------------------------------------------
    if report.type == "enum_ranking":
        _require_run_status(report.status)
        if not report.kind or not ENUM_IDENTIFIER_RE.fullmatch(report.kind):
            raise ValidationError(
                field="kind",
                message=(
                    f"enum_ranking kind must match "
                    f"{ENUM_IDENTIFIER_RE.pattern}."
                ),
            )
    elif report.type == "tag_ranking":
        _require_run_status(report.status)
    elif report.type == "case_trend":
        if not report.case_path.strip():
            raise ValidationError(
                field="case_path",
                message="case_trend case_path must not be empty.",
            )
    elif report.type == "tag_inventory":
        if not report.tag.strip():
            raise ValidationError(
                field="tag",
                message="tag_inventory tag must not be empty.",
            )
        if not report.scope.strip():
            raise ValidationError(
                field="scope",
                message="tag_inventory scope must not be empty.",
            )

    # -- Data-source shape exclusivity -----------------------------------
    if report.type in RUN_SET_TYPES:
        if report.scope:
            raise ValidationError(
                field="scope",
                message="scope must be empty for run-set report types.",
            )
        if report.tag:
            raise ValidationError(
                field="tag",
                message="tag must be empty for run-set report types.",
            )
        if len(report.run_paths) > 10:
            raise ValidationError(
                field="run_paths",
                message="A report may reference at most 10 runs.",
            )
        seen: set[str] = set()
        for i, p in enumerate(report.run_paths):
            if not p:
                raise ValidationError(
                    field=f"run_paths[{i}]",
                    message="run path must not be empty.",
                )
            if p in seen:
                raise ValidationError(
                    field=f"run_paths[{i}]",
                    message=f"Duplicate run path: {p!r}.",
                )
            seen.add(p)
    else:  # FOLDER_TYPES
        if report.run_paths:
            raise ValidationError(
                field="run_paths",
                message="run_paths must be empty for folder report types.",
            )


def _require_run_status(status: str) -> None:
    if status not in RUN_RESULTS:
        raise ValidationError(
            field="status",
            message=(
                f"Invalid status: {status!r}. "
                f"Must be one of {list(RUN_RESULTS)}."
            ),
        )

File: js/src/lib/report.ts
import { parse as parseYaml, stringify as stringifyYaml } from "yaml";
import { RUN_RESULTS } from "./run";

export const RUN_SET_TYPES = ["enum_ranking", "tag_ranking", "case_trend"] as const;
export const REPORT_TYPES = [...RUN_SET_TYPES, "tag_inventory"] as const;
export type ReportType = (typeof REPORT_TYPES)[number];
export const FOLDER_TYPES = new Set(["tag_inventory"]);

export type Report = {
  type: string;
  title: string;
  created_at: string;
  run_paths: string[];
  scope: string;
  status: string;
  kind: string;
  case_path: string;
  tag: string;
};

export class ReportParseError extends Error {
  constructor(message: string, readonly line = 0, readonly column = 0) { super(message); }
}
export class ReportValidationError extends Error {
  constructor(readonly field: string, message: string) { super(message); }
}
export class ReportConflictError extends Error {}

export function emptyReport(): Report { return { type: "", title: "", created_at: "", run_paths: [], scope: "", status: "", kind: "", case_path: "", tag: "" }; }

export function parseReport(source: string): Report {
  let payload: unknown;
  try { payload = parseYaml(source); }
  catch (error) {
    const mark = (error as { linePos?: Array<{ line: number; col: number }> }).linePos?.[0];
    throw new ReportParseError((error as Error).message, mark?.line ?? 0, mark?.col ?? 0);
  }
  if (!payload || typeof payload !== "object" || Array.isArray(payload)) throw new ReportParseError(`Report file root must be a YAML mapping; got ${Array.isArray(payload) ? "list" : typeof payload}.`);
  const record = payload as Record<string, unknown>;
  const paths = record.run_paths;
  if (paths !== undefined && !Array.isArray(paths)) throw new ReportParseError("Report field 'run_paths' must be a list.");
  return { type: typeof record.type === "string" ? record.type : String(record.type ?? ""), title: typeof record.title === "string" ? record.title : String(record.title ?? ""), created_at: typeof record.created_at === "string" ? record.created_at : String(record.created_at ?? ""), run_paths: (paths as unknown[] | undefined)?.map((value) => String(value)) ?? [], scope: typeof record.scope === "string" ? record.scope : String(record.scope ?? ""), status: typeof record.status === "string" ? record.status : String(record.status ?? ""), kind: typeof record.kind === "string" ? record.kind : String(record.kind ?? ""), case_path: typeof record.case_path === "string" ? record.case_path : String(record.case_path ?? ""), tag: typeof record.tag === "string" ? record.tag : String(record.tag ?? "") };
}

export function validateReport(report: Report): void {
  if (!(REPORT_TYPES as readonly string[]).includes(report.type)) throw new ReportValidationError("type", `Invalid report type: '${report.type}'.`);
  if (!report.title.trim()) throw new ReportValidationError("title", "Report title must not be empty.");
  if (/\r|\n/.test(report.title)) throw new ReportValidationError("title", "Report title must be single-line.");
  if (!report.created_at.trim()) throw new ReportValidationError("created_at", "created_at must not be empty.");
  if (/\r|\n/.test(report.created_at)) throw new ReportValidationError("created_at", "created_at must be single-line.");
  if (report.type === "enum_ranking") {
    if (!(RUN_RESULTS as readonly string[]).includes(report.status)) throw new ReportValidationError("status", `Invalid status: '${report.status}'.`);
    if (!/^[A-Za-z_][A-Za-z0-9_]*$/.test(report.kind)) throw new ReportValidationError("kind", "enum_ranking kind is invalid.");
  } else if (report.type === "tag_ranking") {
    if (!(RUN_RESULTS as readonly string[]).includes(report.status)) throw new ReportValidationError("status", `Invalid status: '${report.status}'.`);
  } else if (report.type === "case_trend") {
    if (!report.case_path.trim()) throw new ReportValidationError("case_path", "case_trend case_path must not be empty.");
  } else if (report.type === "tag_inventory") {
    if (!report.tag.trim()) throw new ReportValidationError("tag", "tag_inventory tag must not be empty.");
    if (!report.scope.trim()) throw new ReportValidationError("scope", "tag_inventory scope must not be empty.");
  }
  if (RUN_SET_TYPES.includes(report.type as (typeof RUN_SET_TYPES)[number])) {
    if (report.scope) throw new ReportValidationError("scope", "scope must be empty for run-set report types.");
    if (report.tag) throw new ReportValidationError("tag", "tag must be empty for run-set report types.");
    if (report.run_paths.length > 10) throw new ReportValidationError("run_paths", "A report may reference at most 10 runs.");
    const seen = new Set<string>();
    for (const [index, path] of report.run_paths.entries()) { if (!path) throw new ReportValidationError(`run_paths[${index}]`, "run path must not be empty."); if (seen.has(path)) throw new ReportValidationError(`run_paths[${index}]`, `Duplicate run path: '${path}'.`); seen.add(path); }
  } else if (report.run_paths.length) throw new ReportValidationError("run_paths", "run_paths must be empty for folder report types.");
}

export function reportToPersisted(report: Report): Record<string, unknown> {
  const out: Record<string, unknown> = { type: report.type, title: report.title, created_at: report.created_at };
  if (report.type === "enum_ranking") Object.assign(out, { status: report.status, kind: report.kind, run_paths: report.run_paths });
  else if (report.type === "tag_ranking") Object.assign(out, { status: report.status, run_paths: report.run_paths });
  else if (report.type === "case_trend") Object.assign(out, { case_path: report.case_path, run_paths: report.run_paths });
  else if (report.type === "tag_inventory") Object.assign(out, { tag: report.tag, scope: report.scope });
  return out;
}

export function serializeReport(report: Report): string { validateReport(report); return stringifyYaml(reportToPersisted(report), { sortMapEntries: false, lineWidth: 0 }); }
